system-contracts/test/NonceHolder.spec.ts

```diff
@@ -0,0 +1,248 @@
+      await expect(nonceHolder.connect(systemAccount).incrementMinNonceIfEquals(2222222)).to.be.rejectedWith(
+        "Incorrect nonce"
+      );
+    });
+
+    it("should increment minNonce if equals to expected", async () => {
+      const expectedNonce = await nonceHolder.getMinNonce(systemAccount.address);
+      await nonceHolder.connect(systemAccount).incrementMinNonceIfEquals(expectedNonce);
+      const result = await nonceHolder.getMinNonce(systemAccount.address);
+      expect(result).to.equal(expectedNonce.add(1));
+    });
+  });
+
+  describe("incrementDeploymentNonce", async () => {
+    it("should revert Only the contract deployer can increment the deployment nonce", async () => {
+      await expect(nonceHolder.incrementDeploymentNonce(deployerAccount.address)).to.be.rejectedWith(
+        "Only the contract deployer can increment the deployment nonce"
+      );
+    });
+
+    it("should increment deployment nonce", async () => {
+      const nonceBefore = await nonceHolder.getDeploymentNonce(wallet.address);
+      const rawNonceBefore = await nonceHolder.getRawNonce(wallet.address);
+      await nonceHolder.connect(deployerAccount).incrementDeploymentNonce(wallet.address);
+      const nonceAfter = await nonceHolder.getDeploymentNonce(wallet.address);
+      const rawNonceAfter = await nonceHolder.getRawNonce(wallet.address);
+
+      expect(nonceAfter).to.equal(nonceBefore.add(BigNumber.from(1)));
+      expect(rawNonceAfter).to.equal(rawNonceBefore.add(BigNumber.from(2).pow(128)));
+    });
+  });
+
+  describe("setValueUnderNonce and getValueUnderNonce", async () => {
+    it("should revert Nonce value cannot be set to 0", async () => {
+      const accountInfo = [1, 0];
+      const encodedAccountInfo = ethers.utils.defaultAbiCoder.encode(["tuple(uint8, uint8)"], [accountInfo]);
+      await setResult("ContractDeployer", "getAccountInfo", [systemAccount.address], {
+        failure: false,
+        returnData: encodedAccountInfo,
+      });
+      await expect(nonceHolder.connect(systemAccount).setValueUnderNonce(124, 0)).to.be.rejectedWith(
+        "Nonce value cannot be set to 0"
+      );
+    });
+
+    it("should revert Previous nonce has not been used", async () => {
+      const accountInfo = [1, 0];
+      const encodedAccountInfo = ethers.utils.defaultAbiCoder.encode(["tuple(uint8, uint8)"], [accountInfo]);
+      await setResult("ContractDeployer", "getAccountInfo", [systemAccount.address], {
+        failure: false,
+        returnData: encodedAccountInfo,
+      });
+      await expect(nonceHolder.connect(systemAccount).setValueUnderNonce(443, 111)).to.be.rejectedWith(
+        "Previous nonce has not been used"
+      );
+    });
+
+    it("should emit ValueSetUnderNonce event", async () => {
+      const expectedNonce = await nonceHolder.getMinNonce(systemAccount.address);
+      await nonceHolder.connect(systemAccount).incrementMinNonceIfEquals(expectedNonce);
```

Comment:
Is this call needed? Same in the next test case

Reply:
The reason is for each run of the test to use a fresh nonce when running on the same node instance. Changed it to increasing now.
But kept it the same in the test below, let me know if you think it should be changed